File: auth/src/main/java/com/example/rest/RegistrationController.java
package com.example.rest;

import com.example.dto.RegistrationRequest;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.userdetails.User;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.provisioning.UserDetailsManager;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.web.bind.annotation.*;


@RestController
@Slf4j
public class RegistrationController {

    private final UserDetailsManager users;
    private final PasswordEncoder encoder;


    public RegistrationController(UserDetailsManager users, PasswordEncoder encoder) {
        this.users = users;
        this.encoder = encoder;
    }

    @PostMapping("/register")
    public ResponseEntity<String> register(@RequestBody RegistrationRequest request) {
        log.info("User registration process established for user {}", request.getUsername());
        if (request.getUsername() == null || request.getPassword() == null) {
            return ResponseEntity.badRequest().body("Username and password must not be null");
        }

        if (users.userExists(request.getUsername())) {
            return ResponseEntity.badRequest().body("User already exists");
        }

        UserDetails user = User.withUsername(request.getUsername())
                .password(encoder.encode(request.getPassword()))
                .roles("USER")
                .build();

        users.createUser(user);
        log.info("User registered: {}", user.getUsername());
        return ResponseEntity.ok("User registered");
    }

}

File: auth/src/main/kotlin/com/example/rest/RegistrationController.kt
package com.example.rest

import com.example.dto.RegistrationRequest
import org.slf4j.LoggerFactory
import org.springframework.http.ResponseEntity
import org.springframework.security.core.userdetails.User
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.security.provisioning.UserDetailsManager
import org.springframework.web.bind.annotation.*

@RestController
class RegistrationController(
        private val users: UserDetailsManager,
        private val encoder: PasswordEncoder
) {

    private val log = LoggerFactory.getLogger(RegistrationController::class.java)

    @PostMapping("/register")
    fun register(@RequestBody request: RegistrationRequest): ResponseEntity<String> {
        log.info("User registration process established for user {}", request.username)

        if (request.username.isNullOrBlank() || request.password.isNullOrBlank()) {
            return ResponseEntity.badRequest().body("Username and password must not be null")
        }

        if (users.userExists(request.username)) {
            return ResponseEntity.badRequest().body("User already exists")
        }

        val user = User.withUsername(request.username)
                .password(encoder.encode(request.password))
                .roles("USER")
                .build()

        users.createUser(user)
        log.info("User registered: {}", user.username)
        return ResponseEntity.ok("User registered")
    }
}
